benchmarks/helics/BenchmarkFederate.hpp

```diff
@@ -0,0 +1,320 @@
+/*
+Copyright (c) 2017-2020,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC.  See
+the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+#pragma once
+
+#include "helics/application_api/CombinationFederate.hpp"
+#include "helics/core/ActionMessage.hpp"
+#include "helics/core/helicsCLI11.hpp"
+#include "helics/helics-config.h"
+#include "helics_benchmark_util.h"
+
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+using namespace helics;
+/** class implementing common functionality for benchmarks */
+class BenchmarkFederate {
+  public:
+    enum class OutputFormat {
+        PLAIN_TEXT,
+        //          JSON
+    };
+
+  public:
+    // getters and setters for parameters
+    /** sets the delta time parameter
+     * @param dt the delta time to set
+     */
+    void setDeltaTime(helics::Time dt) { deltaTime = dt; }
+    /** gets the delta time parameter*/
+    helics::Time getDeltaTime() { return deltaTime; }
+    /** sets the final time parameter
+     * @param t the final/stop time*/
+    void setFinalTime(helics::Time t) { finalTime = t; }
+    /** gets the final time parameter*/
+    helics::Time getFinalTime() { return finalTime; }
+
+    /** sets the index parameter
+     * @param i the index
+     */
+    void setIndex(int i) { index = i; }
+    /** gets the index parameter*/
+    int getIndex() { return index; }
+    /** sets the max index parameter
+     * @param i the max index
+     */
+    void setMaxIndex(int i) { maxIndex = i; }
+    /** gets the max index parameter*/
+    int getMaxIndex() { return maxIndex; }
+
+    // functions for setting callbacks
+    /** sets a callback function to call immediately after doMainLoop() returns, but before
+     * the helics finalize() call
+     * @param cb a function that takes no arguments and returns void
+     */
+    void setBeforeFinalizeCallback(std::function<void()> cb = {})
+    {
+        callBeforeFinalize = std::move(cb);
+    }
+    /** sets a callback function to all after the helics finalize() call completes
+     * @param cb a function that takes no arguments and returns void
+     */
+    void setAfterFinalizeCallback(std::function<void()> cb = {})
+    {
+        callAfterFinalize = std::move(cb);
+    }
+
+    /** sets the output format to use when printing results
+     * @param f format to print results in
+     */
+    void setOutputFormat(OutputFormat f) { result_format = f; }
+
+    // protected to give derived classes more control
+  protected:
+    class Result {
+      public:
+        std::string name;
+        std::string key;
+        std::string value;
+    };
+    std::vector<Result> results; //!< a vector of output results to print
+
+    OutputFormat result_format{OutputFormat::PLAIN_TEXT}; //!< output format for printing results
```

Comment:
For packing purposes this variable could be put next to the bool variables at the end of the member list.  Doubt it matters much, but it is a benchmark.

Reply:
How will it help packing if OutputFormat isn't a bool?

Reply:
most data types are 8 byte aligned on a 64 bit system.  So after result_format you have a 4 byte gap.  

right now after the bools you have a 6 byte gap,  so moving result_format would result in an overall size reduction of 8 bytes.  

Not critical but unless there is a good reason not to (like important stuff being in the first 64 bytes) I try to minimize the gaps and structure sizes.  In this case it may make sense to move the bools or leave them separate for readability.

Reply:
Ah okay, so filling in the gap the bools are leaving open. I moved the bools up to below result_format, and rearranged the callbacks/fedptr/params -- if the size goes above 64bytes it would be nice to keep those ones together since could be accessed in the main loop.